Snowflake.Data/Core/ArrowResultChunk.cs

```diff
@@ -123,106 +164,135 @@ public object ExtractCell(int columnIndex, SFDataType srcType, long scale)
-                        case ArrowTypeId.Decimal128:
-                            return ((Decimal128Array)column).GetValue(_currentRecordIndex);
+                                return _long[columnIndex][_currentRecordIndex] / (decimal)s_powersOf10[scale];
+                        
+                        case Decimal128Array array:
+                            return array.GetValue(_currentRecordIndex);
                     }
                     break;
 
                 case SFDataType.BOOLEAN:
                     return ((BooleanArray)column).GetValue(_currentRecordIndex);
+
                 case SFDataType.REAL:
                     // Snowflake data types that are floating-point numbers will fall in this category
                     // e.g. FLOAT/REAL/DOUBLE
-                    return ((DoubleArray)column).GetValue(_currentRecordIndex);
+                    if (_double[columnIndex] == null)
+                        _double[columnIndex] = ((DoubleArray)column).Values.ToArray();
+                    return _double[columnIndex][_currentRecordIndex];
+
                 case SFDataType.TEXT:
                 case SFDataType.ARRAY:
                 case SFDataType.VARIANT:
                 case SFDataType.OBJECT:
-                    return ((StringArray)column).GetString(_currentRecordIndex);
+                    if (_byte[columnIndex] == null || _int[columnIndex] == null)
+                    {
+                        _byte[columnIndex] = ((StringArray)column).Values.ToArray();
+                        _int[columnIndex] = ((StringArray)column).ValueOffsets.ToArray();
+                    }
+                    return StringArray.DefaultEncoding.GetString(
+                        _byte[columnIndex], 
+                        _int[columnIndex][_currentRecordIndex], 
+                        _int[columnIndex][_currentRecordIndex + 1] - _int[columnIndex][_currentRecordIndex]);
+                    
                 case SFDataType.BINARY:
                     return ((BinaryArray)column).GetBytes(_currentRecordIndex).ToArray();
+                
                 case SFDataType.DATE:
-                    return ((Date32Array)column).GetDateTime(_currentRecordIndex);
+                    if (_int[columnIndex] == null)
+                        _int[columnIndex] = ((Date32Array)column).Values.ToArray();
+                    return SFDataConverter.UnixEpoch.AddTicks(_int[columnIndex][_currentRecordIndex] * TicksPerDay);
+                
                 case SFDataType.TIME:
                 {
-                    var value = column.Data.DataType.TypeId == ArrowTypeId.Int32
-                        ? ((Int32Array)column).GetValue(_currentRecordIndex)
-                        : ((Int64Array)column).GetValue(_currentRecordIndex);
+                    var value = column.GetType() == typeof(Int32Array)
+                        ? ((Int32Array)column).Values[_currentRecordIndex]
```

Comment:
I think it should be also replaced by cached values because Values property is not an array, it is implemented as
`public ReadOnlySpan<T> Values => ValueBuffer.Span.CastTo<T>().Slice(Offset, Length);`
so it is not efficient.

Reply:
Done